### DISCUSSION TITLE
improving String Formatting Components

### DISCUSSION THREAD

**Discussion by Saevon** [FakeFishGames/Barotrauma#13118] [Ideas and suggestions]
I'll often see people use 2 `text displays` with hard coded values rather then just one `text display` and `concat components`.

## Suggestion 1

If there was a `format component` which supported more general string formatting this would be much easier!
- 5 inputs, 1 output
- Choose a "format" value using a simple format: e.g. `{input_1}` (editable in-game)
- It wouldn't need parsing since 5 hard-coded "search and replace" could work just fine!

e.g. `Reactor Condition:  {input1} %`

Similarly the `concat component` isn't great, as it can only handle 2 things

## Suggestion 2
`Concat Components` should have 6 inputs
- Reads ALL inputs that aren't NULL and joins them with the separator
- This would actually replace the `color component` as its just a `concat` with 4 inputs and `,` separator

Note: similarly `and` / `or` / `add` / `mult` should have multiple inputs as well)
- Also `sub` could if the first was `value_in` and the remaining were `sub_in_1` / `sub_in_2` / `sub_in_3` etc

## Version 3:

See #8699 which adds an "output format" directly to many components instead

**Comment by Arctodites** [comment]
A 'FORMAT' component could be useful, especially since (to my understanding) the REGEX component only permits us to work with one capture group.  I don't know how that compares "under the hood," w.r.t. computational load on the game engine, but I'd also like for AND and OR components to have at least three input terminals.

I suspect another reason builders often use two text displays instead of the concat component, is that the text display itself is fairly arcane to use: depending on the scale of the item itself, the scale of the text's font, the specific alphanumeric or symbol characters used, and the length of the string to be printed, it might appear as solid text or as a scrolling message, and it might be aligned with nearby texts or it might not.  Using two hard-coded display units allows slightly more text to be displayed at the same time, and reduces the uncertainty of how it will appear.  What we really need is a text display item that can display longer messages all at once -- preferably an extensible object (that can be resized like submarine shells, wall textures, etc. using control points) or one that can be seamlessly tiled and chained together with other units into a larger, synchronised display.  Being able to choose in the sub editor between left, centre, and right justifications for our text would even be the icing on the cake.
